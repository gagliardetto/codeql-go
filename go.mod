module github.com/github/codeql-go

go 1.14

require (
	github.com/gin-gonic/gin v1.6.2
	golang.org/x/mod v0.2.0
	golang.org/x/tools v0.0.0-20200302225559-9b52d559c609
)
