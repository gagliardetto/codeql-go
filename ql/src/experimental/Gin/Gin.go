package main

func main() {}

// gin
func ginHandler(c *gin.Context) {
	{
		header := c.GetHeader("key")
		_ = header
	}
	{
		params := c.QueryArray("key")
		_ = params
	}
	{
		keys := c.Keys
		_ = keys
	}

	c.BindYAML
	c.BindXML
	c.BindWith
	c.BindUri
	c.BindQuery
	c.MustBindWith
	c.BindJSON
	c.Bind
	c.ShouldBind
	c.ShouldBindBodyWith
	c.ShouldBindJSON
	c.ShouldBindQuery
	c.ShouldBindUri
	c.ShouldBindWith
	c.ShouldBindXML
	c.ShouldBindYAML
	//c.SaveUploadedFile
}
