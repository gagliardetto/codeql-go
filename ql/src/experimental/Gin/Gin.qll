/**
 * Provides classes for working with untrusted flow sources from the `github.com/gin-gonic/gin` package.
 */

import go

///::: "github.com/gin-gonic/gin".Context
///: type:{"github.com/gin-gonic/gin".Context}
///: type:{"github.com/gin-gonic/gin".Context}/methods/Copy/out/0/type:{"github.com/gin-gonic/gin".Context}
///: type:{"github.com/gin-gonic/gin".Context}/methods/Handler/out/0/type:{"github.com/gin-gonic/gin".HandlerFunc}/methods/HandlerFunc/in/0/type:{"github.com/gin-gonic/gin".Context}

// ------------------------------------------------------------
// "github.com/gin-gonic/gin".Context sources:
///: kind:struct
class GithubComGinGonicGinContextSource extends UntrustedFlowSource::Range, DataFlow::Node {
  GithubComGinGonicGinContextSource() {
    exists(string packagePath, string typeName |
      packagePath = "github.com/gin-gonic/gin" and
      typeName = "Context"
    |
      // Any read of a variable of this type:
      exists(DataFlow::ReadNode read |
        read.getType().hasQualifiedName(packagePath, typeName)
      |
        this = read
      )
      or
      // Method calls:
      exists(DataFlow::MethodCallNode call, string methodName, int resultIndex |
        call.getTarget().hasQualifiedName(packagePath, typeName, methodName) and
        (
          methodName = "GetHeader"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
          or
          methodName = "QueryArray"
            and
            (
              ///:out i:0 kind:slice type:[]string
              resultIndex = 0
            )
          or
          methodName = "Query"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
          or
          methodName = "PostFormArray"
            and
            (
              ///:out i:0 kind:slice type:[]string
              resultIndex = 0
            )
          or
          methodName = "PostForm"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
          or
          methodName = "Param"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
          or
          methodName = "NegotiateFormat"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
          or
          methodName = "HandlerNames"
            and
            (
              ///:out i:0 kind:slice type:[]string
              resultIndex = 0
            )
          or
          methodName = "HandlerName"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
          or
          methodName = "GetStringSlice"
            and
            (
              ///:out i:0 kind:slice type:[]string
              resultIndex = 0
            )
          or
          methodName = "GetString"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
          or
          methodName = "GetRawData"
            and
            (
              ///:out i:0 kind:slice type:[]uint8
              resultIndex = 0
                or
              ///:out i:1 kind:interface type:error
              resultIndex = 1
            )
          or
          methodName = "ClientIP"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
          or
          methodName = "ContentType"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
          or
          methodName = "Cookie"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
                or
              ///:out i:1 kind:interface type:error
              resultIndex = 1
            )
          or
          methodName = "GetQueryArray"
            and
            (
              ///:out i:0 kind:slice type:[]string
              resultIndex = 0
                or
              ///:out i:1 kind:bool type:bool
              resultIndex = 1
            )
          or
          methodName = "GetQuery"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
                or
              ///:out i:1 kind:bool type:bool
              resultIndex = 1
            )
          or
          methodName = "GetPostFormArray"
            and
            (
              ///:out i:0 kind:slice type:[]string
              resultIndex = 0
                or
              ///:out i:1 kind:bool type:bool
              resultIndex = 1
            )
          or
          methodName = "GetPostForm"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
                or
              ///:out i:1 kind:bool type:bool
              resultIndex = 1
            )
          or
          methodName = "DefaultPostForm"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
          or
          methodName = "DefaultQuery"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
          or
          methodName = "Err"
            and
            (
              ///:out i:0 kind:interface type:error
              resultIndex = 0
            )
          or
          methodName = "Error"
            and
            (
              ///:out i:0 kind:ptr type:"github.com/gin-gonic/gin".Error
              resultIndex = 0
            )
          or
          methodName = "FormFile"
            and
            (
              ///:out i:0 kind:ptr type:"mime/multipart".FileHeader
              resultIndex = 0
                or
              ///:out i:1 kind:interface type:error
              resultIndex = 1
            )
          or
          methodName = "Get"
            and
            (
              ///:out i:0 kind:interface type:interface {}
              resultIndex = 0
                or
              ///:out i:1 kind:bool type:bool
              resultIndex = 1
            )
          or
          methodName = "GetBool"
            and
            (
              ///:out i:0 kind:bool type:bool
              resultIndex = 0
            )
          or
          methodName = "GetDuration"
            and
            (
              ///:out i:0 kind:int64 type:"time".Duration
              resultIndex = 0
            )
          or
          methodName = "GetFloat64"
            and
            (
              ///:out i:0 kind:float64 type:float64
              resultIndex = 0
            )
          or
          methodName = "Done"
            and
            (
              ///:out i:0 kind:chan type:<-chan struct {}
              resultIndex = 0
            )
          or
          methodName = "GetInt"
            and
            (
              ///:out i:0 kind:int type:int
              resultIndex = 0
            )
          or
          methodName = "GetInt64"
            and
            (
              ///:out i:0 kind:int64 type:int64
              resultIndex = 0
            )
          or
          methodName = "Deadline"
            and
            (
              ///:out i:0 kind:struct type:"time".Time
              resultIndex = 0
                or
              ///:out i:1 kind:bool type:bool
              resultIndex = 1
            )
          or
          methodName = "GetPostFormMap"
            and
            (
              ///:out i:0 kind:map type:map[string]string
              resultIndex = 0
                or
              ///:out i:1 kind:bool type:bool
              resultIndex = 1
            )
          or
          methodName = "Copy"
            and
            (
              ///:out i:0 kind:ptr type:"github.com/gin-gonic/gin".Context
              resultIndex = 0
            )
          or
          methodName = "GetQueryMap"
            and
            (
              ///:out i:0 kind:map type:map[string]string
              resultIndex = 0
                or
              ///:out i:1 kind:bool type:bool
              resultIndex = 1
            )
          or
          methodName = "BindYAML"
            and
            (
              ///:out i:0 kind:interface type:error
              resultIndex = 0
            )
          or
          methodName = "BindXML"
            and
            (
              ///:out i:0 kind:interface type:error
              resultIndex = 0
            )
          or
          methodName = "GetStringMap"
            and
            (
              ///:out i:0 kind:map type:map[string]interface {}
              resultIndex = 0
            )
          or
          methodName = "GetStringMapString"
            and
            (
              ///:out i:0 kind:map type:map[string]string
              resultIndex = 0
            )
          or
          methodName = "GetStringMapStringSlice"
            and
            (
              ///:out i:0 kind:map type:map[string][]string
              resultIndex = 0
            )
          or
          methodName = "BindWith"
            and
            (
              ///:out i:0 kind:interface type:error
              resultIndex = 0
            )
          or
          methodName = "GetTime"
            and
            (
              ///:out i:0 kind:struct type:"time".Time
              resultIndex = 0
            )
          or
          methodName = "Handler"
            and
            (
              ///:out i:0 kind:func type:"github.com/gin-gonic/gin".HandlerFunc
              resultIndex = 0
            )
          or
          methodName = "BindUri"
            and
            (
              ///:out i:0 kind:interface type:error
              resultIndex = 0
            )
          or
          methodName = "BindQuery"
            and
            (
              ///:out i:0 kind:interface type:error
              resultIndex = 0
            )
          or
          methodName = "IsAborted"
            and
            (
              ///:out i:0 kind:bool type:bool
              resultIndex = 0
            )
          or
          methodName = "IsWebsocket"
            and
            (
              ///:out i:0 kind:bool type:bool
              resultIndex = 0
            )
          or
          methodName = "MultipartForm"
            and
            (
              ///:out i:0 kind:ptr type:"mime/multipart".Form
              resultIndex = 0
                or
              ///:out i:1 kind:interface type:error
              resultIndex = 1
            )
          or
          methodName = "MustBindWith"
            and
            (
              ///:out i:0 kind:interface type:error
              resultIndex = 0
            )
          or
          methodName = "MustGet"
            and
            (
              ///:out i:0 kind:interface type:interface {}
              resultIndex = 0
            )
          or
          methodName = "BindJSON"
            and
            (
              ///:out i:0 kind:interface type:error
              resultIndex = 0
            )
          or
          methodName = "Bind"
            and
            (
              ///:out i:0 kind:interface type:error
              resultIndex = 0
            )
          or
          methodName = "PostFormMap"
            and
            (
              ///:out i:0 kind:map type:map[string]string
              resultIndex = 0
            )
          or
          methodName = "AbortWithError"
            and
            (
              ///:out i:0 kind:ptr type:"github.com/gin-gonic/gin".Error
              resultIndex = 0
            )
          or
          methodName = "QueryMap"
            and
            (
              ///:out i:0 kind:map type:map[string]string
              resultIndex = 0
            )
          or
          methodName = "SaveUploadedFile"
            and
            (
              ///:out i:0 kind:interface type:error
              resultIndex = 0
            )
          or
          methodName = "ShouldBind"
            and
            (
              ///:out i:0 kind:interface type:error
              resultIndex = 0
            )
          or
          methodName = "ShouldBindBodyWith"
            and
            (
              ///:out i:0 kind:interface type:error
              resultIndex = 0
            )
          or
          methodName = "ShouldBindJSON"
            and
            (
              ///:out i:0 kind:interface type:error
              resultIndex = 0
            )
          or
          methodName = "ShouldBindQuery"
            and
            (
              ///:out i:0 kind:interface type:error
              resultIndex = 0
            )
          or
          methodName = "ShouldBindUri"
            and
            (
              ///:out i:0 kind:interface type:error
              resultIndex = 0
            )
          or
          methodName = "ShouldBindWith"
            and
            (
              ///:out i:0 kind:interface type:error
              resultIndex = 0
            )
          or
          methodName = "ShouldBindXML"
            and
            (
              ///:out i:0 kind:interface type:error
              resultIndex = 0
            )
          or
          methodName = "ShouldBindYAML"
            and
            (
              ///:out i:0 kind:interface type:error
              resultIndex = 0
            )
          or
          methodName = "Stream"
            and
            (
              ///:out i:0 kind:bool type:bool
              resultIndex = 0
            )
          or
          methodName = "Value"
            and
            (
              ///:out i:0 kind:interface type:interface {}
              resultIndex = 0
            )
        )
      |
        resultIndex = 0 and this = call.getResult()
        or
        this = call.getResult(resultIndex)
      )
      or
      // Field reads:
      exists(DataFlow::Field fld, string fieldName |
        (
          ///: kind:slice type:[]string
          fieldName = "Accepted"
            or
          ///: kind:ptr type:"net/http".Request
          fieldName = "Request"
            or
          ///: kind:slice type:"github.com/gin-gonic/gin".Params
          fieldName = "Params"
            or
          ///: kind:interface type:"github.com/gin-gonic/gin".ResponseWriter
          fieldName = "Writer"
            or
          ///: kind:slice type:"github.com/gin-gonic/gin".errorMsgs
          fieldName = "Errors"
            or
          ///: kind:map type:map[string]interface {}
          fieldName = "Keys"
        )
      |
        fld.hasQualifiedName(packagePath, typeName, fieldName) and 
        this = fld.getARead()
      )
    )
  }
}

///::: "github.com/gin-gonic/gin".Params
///: type:{"github.com/gin-gonic/gin".Context}/fields/Params/type:{"github.com/gin-gonic/gin".Params}

// ------------------------------------------------------------
// "github.com/gin-gonic/gin".Params sources:
///: kind:slice
class GithubComGinGonicGinParamsSource extends UntrustedFlowSource::Range, DataFlow::Node {
  GithubComGinGonicGinParamsSource() {
    exists(string packagePath, string typeName |
      packagePath = "github.com/gin-gonic/gin" and
      typeName = "Params"
    |
      // Any read of a variable of this type:
      exists(DataFlow::ReadNode read |
        read.getType().hasQualifiedName(packagePath, typeName)
      |
        this = read
      )
      or
      // Method calls:
      exists(DataFlow::MethodCallNode call, string methodName, int resultIndex |
        call.getTarget().hasQualifiedName(packagePath, typeName, methodName) and
        (
          methodName = "ByName"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
          or
          methodName = "Get"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
                or
              ///:out i:1 kind:bool type:bool
              resultIndex = 1
            )
        )
      |
        resultIndex = 0 and this = call.getResult()
        or
        this = call.getResult(resultIndex)
      )
    )
  }
}

///::: "github.com/gin-gonic/gin".Param
///: type:{"github.com/gin-gonic/gin".Context}/fields/Params/type:{"github.com/gin-gonic/gin".Params}/array/type:{"github.com/gin-gonic/gin".Param}

// ------------------------------------------------------------
// "github.com/gin-gonic/gin".Param sources:
///: kind:struct
class GithubComGinGonicGinParamSource extends UntrustedFlowSource::Range, DataFlow::Node {
  GithubComGinGonicGinParamSource() {
    exists(string packagePath, string typeName |
      packagePath = "github.com/gin-gonic/gin" and
      typeName = "Param"
    |
      // Any read of a variable of this type:
      exists(DataFlow::ReadNode read |
        read.getType().hasQualifiedName(packagePath, typeName)
      |
        this = read
      )
      or
      // Field reads:
      exists(DataFlow::Field fld, string fieldName |
        (
          ///: kind:string type:string
          fieldName = "Key"
            or
          ///: kind:string type:string
          fieldName = "Value"
        )
      |
        fld.hasQualifiedName(packagePath, typeName, fieldName) and 
        this = fld.getARead()
      )
    )
  }
}

class GithubComGinGonicGinContextBindSource extends UntrustedFlowSource::Range, DataFlow::Node {
  GithubComGinGonicGinContextBindSource() {
    exists(string packagePath, string typeName |
      packagePath = "github.com/gin-gonic/gin" and
      typeName = "Context"
    |
      exists(DataFlow::MethodCallNode call, string methodName |
        call.getTarget().hasQualifiedName(packagePath, typeName, methodName) and
        (
          methodName = "BindYAML" or
          methodName = "BindXML" or
          methodName = "BindWith" or
          methodName = "BindUri" or
          methodName = "BindQuery" or
          methodName = "MustBindWith" or
          methodName = "BindJSON" or
          methodName = "Bind" or
          methodName = "ShouldBind" or
          methodName = "ShouldBindBodyWith" or
          methodName = "ShouldBindJSON" or
          methodName = "ShouldBindQuery" or
          methodName = "ShouldBindUri" or
          methodName = "ShouldBindWith" or
          methodName = "ShouldBindXML" or
          methodName = "ShouldBindYAML"
        )
      |
        this = call.getArgument(0)
      )
    )
  }
}