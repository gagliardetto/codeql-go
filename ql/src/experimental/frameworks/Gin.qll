/**
 * Provides classes for working with HTTP-related concepts such as requests and responses from the `github.com/gin-gonic/gin` package.
 */

import go

private module Gin {
// "github.com/gin-gonic/gin".Context sources:
///: kind:struct
private class GithubComGinGonicGinContextSource extends UntrustedFlowSource::Range, DataFlow::Node {
  GithubComGinGonicGinContextSource() {
    exists(string packagePath, string typeName |
      packagePath = "github.com/gin-gonic/gin" and
      typeName = "Context"
    |
      // Method calls:
      exists(DataFlow::MethodCallNode call, string methodName, int resultIndex |
        call.getTarget().hasQualifiedName(packagePath, typeName, methodName) and
        (
          methodName = "GetHeader"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
          or
          methodName = "QueryArray"
            and
            (
              ///:out i:0 kind:slice type:[]string
              resultIndex = 0
            )
          or
          methodName = "Query"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
          or
          methodName = "PostFormArray"
            and
            (
              ///:out i:0 kind:slice type:[]string
              resultIndex = 0
            )
          or
          methodName = "PostForm"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
          or
          methodName = "Param"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
          or
          methodName = "GetStringSlice"
            and
            (
              ///:out i:0 kind:slice type:[]string
              resultIndex = 0
            )
          or
          methodName = "GetString"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
          or
          methodName = "GetRawData"
            and
            (
              ///:out i:0 kind:slice type:[]uint8
              resultIndex = 0
            )
          or
          methodName = "ClientIP"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
          or
          methodName = "ContentType"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
          or
          methodName = "Cookie"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
          or
          methodName = "GetQueryArray"
            and
            (
              ///:out i:0 kind:slice type:[]string
              resultIndex = 0
            )
          or
          methodName = "GetQuery"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
          or
          methodName = "GetPostFormArray"
            and
            (
              ///:out i:0 kind:slice type:[]string
              resultIndex = 0
            )
          or
          methodName = "GetPostForm"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
          or
          methodName = "DefaultPostForm"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
          or
          methodName = "DefaultQuery"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
          or
          methodName = "Get"
            and
            (
              ///:out i:0 kind:interface type:interface {}
              resultIndex = 0
            )
          or
          methodName = "GetPostFormMap"
            and
            (
              ///:out i:0 kind:map type:map[string]string
              resultIndex = 0
            )
          or
          methodName = "GetQueryMap"
            and
            (
              ///:out i:0 kind:map type:map[string]string
              resultIndex = 0
            )
          or
          methodName = "GetStringMap"
            and
            (
              ///:out i:0 kind:map type:map[string]interface {}
              resultIndex = 0
            )
          or
          methodName = "GetStringMapString"
            and
            (
              ///:out i:0 kind:map type:map[string]string
              resultIndex = 0
            )
          or
          methodName = "GetStringMapStringSlice"
            and
            (
              ///:out i:0 kind:map type:map[string][]string
              resultIndex = 0
            )
          or
          methodName = "MustGet"
            and
            (
              ///:out i:0 kind:interface type:interface {}
              resultIndex = 0
            )
          or
          methodName = "PostFormMap"
            and
            (
              ///:out i:0 kind:map type:map[string]string
              resultIndex = 0
            )
          or
          methodName = "QueryMap"
            and
            (
              ///:out i:0 kind:map type:map[string]string
              resultIndex = 0
            )
        )
      |
        resultIndex = 0 and this = call.getResult()
        or
        this = call.getResult(resultIndex)
      )
      or
      // Field reads:
      exists(DataFlow::Field fld, string fieldName |
        (
          ///: kind:slice type:[]string
          fieldName = "Accepted"
        )
      |
        fld.hasQualifiedName(packagePath, typeName, fieldName) and 
        this = fld.getARead()
      )
    )
  }
}

// "github.com/gin-gonic/gin".Params sources:
///: kind:slice
private class GithubComGinGonicGinParamsSource extends UntrustedFlowSource::Range, DataFlow::Node {
  GithubComGinGonicGinParamsSource() {
    exists(string packagePath, string typeName |
      packagePath = "github.com/gin-gonic/gin" and
      typeName = "Params"
    |
      // Any read of a variable of this type:
      exists(DataFlow::ReadNode read |
        read.getType().hasQualifiedName(packagePath, typeName)
      |
        this = read
      )
      or
      // Method calls:
      exists(DataFlow::MethodCallNode call, string methodName, int resultIndex |
        call.getTarget().hasQualifiedName(packagePath, typeName, methodName) and
        (
          methodName = "ByName"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
          or
          methodName = "Get"
            and
            (
              ///:out i:0 kind:string type:string
              resultIndex = 0
            )
        )
      |
        resultIndex = 0 and this = call.getResult()
        or
        this = call.getResult(resultIndex)
      )
    )
  }
}

///::: "github.com/gin-gonic/gin".Param
///: type:{"github.com/gin-gonic/gin".Context}/fields/Params/type:{"github.com/gin-gonic/gin".Params}/array/type:{"github.com/gin-gonic/gin".Param}

// ------------------------------------------------------------
// "github.com/gin-gonic/gin".Param sources:
///: kind:struct
private class GithubComGinGonicGinParamSource extends UntrustedFlowSource::Range, DataFlow::Node {
  GithubComGinGonicGinParamSource() {
    exists(string packagePath, string typeName |
      packagePath = "github.com/gin-gonic/gin" and
      typeName = "Param"
    |
      // Any read of a variable of this type:
      exists(DataFlow::ReadNode read |
        read.getType().hasQualifiedName(packagePath, typeName)
      |
        this = read
      )
      or
      // Field reads:
      exists(DataFlow::Field fld, string fieldName |
        (
          ///: kind:string type:string
          fieldName = "Key"
            or
          ///: kind:string type:string
          fieldName = "Value"
        )
      |
        fld.hasQualifiedName(packagePath, typeName, fieldName) and 
        this = fld.getARead()
      )
    )
  }
}

private class GithubComGinGonicGinContextBindSource extends UntrustedFlowSource::Range, DataFlow::Node {
  GithubComGinGonicGinContextBindSource() {
    exists(string packagePath, string typeName |
      packagePath = "github.com/gin-gonic/gin" and
      typeName = "Context"
    |
      exists(DataFlow::MethodCallNode call, string methodName |
        call.getTarget().hasQualifiedName(packagePath, typeName, methodName) and
        (
          methodName = "BindYAML" or
          methodName = "BindXML" or
          methodName = "BindWith" or
          methodName = "BindUri" or
          methodName = "BindQuery" or
          methodName = "MustBindWith" or
          methodName = "BindJSON" or
          methodName = "Bind" or
          methodName = "ShouldBind" or
          methodName = "ShouldBindBodyWith" or
          methodName = "ShouldBindJSON" or
          methodName = "ShouldBindQuery" or
          methodName = "ShouldBindUri" or
          methodName = "ShouldBindWith" or
          methodName = "ShouldBindXML" or
          methodName = "ShouldBindYAML"
        )
      |
        this = call.getArgument(0)
      )
    )
  }
}
}