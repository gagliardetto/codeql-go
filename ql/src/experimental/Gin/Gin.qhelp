import go
import experimental.Gin

from UntrustedFlowSource exec
select exec, ""