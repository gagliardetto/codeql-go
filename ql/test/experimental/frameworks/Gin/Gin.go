package main

import (
	"os/exec"

	"github.com/gin-gonic/gin"
	"github.com/gin-gonic/gin/binding"
)

func main() {}

type Person struct {
	Name    string `form:"name"`
	Address string `form:"address"`
}

func use(val string) {
	exec.Command(val)
}

// gin
func ginHandleruse(ctx *gin.Context) {
	{
		val := ctx.GetHeader("key")
		use(val)
	}
	{
		val := ctx.QueryArray("key")
		use(val[0])
	}
	{
		val := ctx.Query("key")
		use(val)
	}
	{
		val := ctx.PostFormArray("key")
		use(val[0])
	}
	{
		val := ctx.PostForm("key")
		use(val)
	}
	{
		val := ctx.Param("key")
		use(val)
	}
	{
		val := ctx.GetStringSlice("key")
		use(val[0])
	}
	{
		val := ctx.GetString("key")
		use(val)
	}
	{
		val, _ := ctx.GetRawData()
		use(string(val))
	}
	{
		val := ctx.ClientIP()
		use(val)
	}
	{
		val := ctx.ContentType()
		use(val)
	}
	{
		val, _ := ctx.Cookie("key")
		use(val)
	}
	{
		val, _ := ctx.GetQueryArray("key")
		use(val[0])
	}
	{
		val, _ := ctx.GetQuery("key")
		use(val)
	}
	{
		val, _ := ctx.GetPostFormArray("key")
		use(val[0])
	}
	{
		val, _ := ctx.GetPostForm("key")
		use(val)
	}
	{
		val := ctx.DefaultPostForm("key", "default-value")
		use(val)
	}
	{
		val := ctx.DefaultQuery("key", "default-value")
		use(val)
	}
	{
		val, _ := ctx.Get("key")
		use(val.(string))
	}
	{
		val, _ := ctx.GetPostFormMap("key")
		use(val["a"])
	}
	{
		val, _ := ctx.GetQueryMap("key")
		use(val["a"])
	}
	{
		val := ctx.GetStringMap("key")
		use(val["a"].(string))
	}
	{
		val := ctx.GetStringMapString("key")
		use(val["a"])
	}
	{
		val := ctx.GetStringMapStringSlice("key")
		use(val["a"][0])
	}
	{
		val := ctx.MustGet("key")
		use(val.(string))
	}
	{
		val := ctx.PostFormMap("key")
		use(val["a"])
	}
	{
		val := ctx.QueryMap("key")
		use(val["a"])
	}

	// fields:

	{
		val := ctx.Accepted
		use(val[0])
	}
	{
		val := ctx.Params
		use(val[0].Value)
	}

	// Params:
	{
		val := ctx.Params[0]
		use(val.Value)
	}
	{
		val := ctx.Params.ByName("name")
		use(val)
	}
	{
		val, _ := ctx.Params.Get("name")
		use(val)
	}

	// Param:
	{
		param := ctx.Params[0]
		key := param.Key
		val := param.Value
		use(key)
		use(val)
	}

	// bind:
	{
		var person Person
		ctx.BindYAML(&person)
		_ = person
	}
	{
		var person Person
		ctx.BindXML(&person)
		use(person.Name)
	}
	{
		var person Person
		ctx.BindWith(&person, binding.YAML)
		use(person.Name)
	}
	{
		var person Person
		ctx.BindUri(&person)
		use(person.Name)
	}
	{
		var person Person
		ctx.BindQuery(&person)
		use(person.Name)
	}
	{
		var person Person
		ctx.MustBindWith(&person, binding.YAML)
		use(person.Name)
	}
	{
		var person Person
		ctx.BindJSON(&person)
		use(person.Name)
	}
	{
		var person Person
		ctx.Bind(&person)
		use(person.Name)
	}
	{
		var person Person
		ctx.ShouldBind(&person)
		use(person.Name)
	}
	{
		var person Person
		ctx.ShouldBindBodyWith(&person, binding.YAML)
		use(person.Name)
	}
	{
		var person Person
		ctx.ShouldBindJSON(&person)
		use(person.Name)
	}
	{
		var person Person
		ctx.ShouldBindQuery(&person)
		use(person.Name)
	}
	{
		var person Person
		ctx.ShouldBindUri(&person)
		use(person.Name)
	}
	{
		var person Person
		ctx.ShouldBindWith(&person, binding.YAML)
		use(person.Name)
	}
	{
		var person Person
		ctx.ShouldBindXML(&person)
		use(person.Name)
	}
	{
		var person Person
		ctx.ShouldBindYAML(&person)
		use(person.Name)
	}
}
