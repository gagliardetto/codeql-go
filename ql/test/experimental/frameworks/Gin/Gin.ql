import go
import experimental.frameworks.Gin

from UntrustedFlowSource exec
select exec, "untrusted source"