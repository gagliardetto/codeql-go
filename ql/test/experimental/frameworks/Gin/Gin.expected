| Gin.go:24:10:24:29 | call to GetHeader | untrusted source |
| Gin.go:28:10:28:30 | call to QueryArray | untrusted source |
| Gin.go:32:10:32:25 | call to Query | untrusted source |
| Gin.go:36:10:36:33 | call to PostFormArray | untrusted source |
| Gin.go:40:10:40:28 | call to PostForm | untrusted source |
| Gin.go:44:10:44:25 | call to Param | untrusted source |
| Gin.go:48:10:48:34 | call to GetStringSlice | untrusted source |
| Gin.go:52:10:52:29 | call to GetString | untrusted source |
| Gin.go:56:3:56:28 | ... := ...[0] | untrusted source |
| Gin.go:60:10:60:23 | call to ClientIP | untrusted source |
| Gin.go:64:10:64:26 | call to ContentType | untrusted source |
| Gin.go:68:3:68:29 | ... := ...[0] | untrusted source |
| Gin.go:72:3:72:36 | ... := ...[0] | untrusted source |
| Gin.go:76:3:76:31 | ... := ...[0] | untrusted source |
| Gin.go:80:3:80:39 | ... := ...[0] | untrusted source |
| Gin.go:84:3:84:34 | ... := ...[0] | untrusted source |
| Gin.go:88:10:88:52 | call to DefaultPostForm | untrusted source |
| Gin.go:92:10:92:49 | call to DefaultQuery | untrusted source |
| Gin.go:96:3:96:26 | ... := ...[0] | untrusted source |
| Gin.go:100:3:100:37 | ... := ...[0] | untrusted source |
| Gin.go:104:3:104:34 | ... := ...[0] | untrusted source |
| Gin.go:108:10:108:32 | call to GetStringMap | untrusted source |
| Gin.go:112:10:112:38 | call to GetStringMapString | untrusted source |
| Gin.go:116:10:116:43 | call to GetStringMapStringSlice | untrusted source |
| Gin.go:120:10:120:27 | call to MustGet | untrusted source |
| Gin.go:124:10:124:31 | call to PostFormMap | untrusted source |
| Gin.go:128:10:128:28 | call to QueryMap | untrusted source |
| Gin.go:135:10:135:21 | selection of Accepted | untrusted source |
| Gin.go:139:10:139:19 | selection of Params | untrusted source |
| Gin.go:140:7:140:9 | val | untrusted source |
| Gin.go:140:7:140:12 | index expression | untrusted source |
| Gin.go:140:7:140:18 | selection of Value | untrusted source |
| Gin.go:145:10:145:19 | selection of Params | untrusted source |
| Gin.go:145:10:145:22 | index expression | untrusted source |
| Gin.go:146:7:146:9 | val | untrusted source |
| Gin.go:146:7:146:15 | selection of Value | untrusted source |
| Gin.go:149:10:149:19 | selection of Params | untrusted source |
| Gin.go:149:10:149:34 | call to ByName | untrusted source |
| Gin.go:153:3:153:34 | ... := ...[0] | untrusted source |
| Gin.go:153:13:153:22 | selection of Params | untrusted source |
| Gin.go:159:12:159:21 | selection of Params | untrusted source |
| Gin.go:159:12:159:24 | index expression | untrusted source |
| Gin.go:160:10:160:14 | param | untrusted source |
| Gin.go:160:10:160:18 | selection of Key | untrusted source |
| Gin.go:161:10:161:14 | param | untrusted source |
| Gin.go:161:10:161:20 | selection of Value | untrusted source |
| Gin.go:169:16:169:22 | &... | untrusted source |
| Gin.go:174:15:174:21 | &... | untrusted source |
| Gin.go:179:16:179:22 | &... | untrusted source |
| Gin.go:184:15:184:21 | &... | untrusted source |
| Gin.go:189:17:189:23 | &... | untrusted source |
| Gin.go:194:20:194:26 | &... | untrusted source |
| Gin.go:199:16:199:22 | &... | untrusted source |
| Gin.go:204:12:204:18 | &... | untrusted source |
| Gin.go:209:18:209:24 | &... | untrusted source |
| Gin.go:214:26:214:32 | &... | untrusted source |
| Gin.go:219:22:219:28 | &... | untrusted source |
| Gin.go:224:23:224:29 | &... | untrusted source |
| Gin.go:229:21:229:27 | &... | untrusted source |
| Gin.go:234:22:234:28 | &... | untrusted source |
| Gin.go:239:21:239:27 | &... | untrusted source |
| Gin.go:244:22:244:28 | &... | untrusted source |
